core/store/src/db/refcount.rs

```diff
@@ -112,27 +112,55 @@ pub(crate) fn encode_negative_refcount(rc: std::num::NonZeroU32) -> [u8; 8] {
 /// Assumes that all provided values with positive reference count have the same
 /// value so that the function is free to pick any of the values.  In build with
 /// debug assertions panics if this is not true.
-pub(crate) fn refcount_merge<'a>(
+pub(crate) fn refcount_full_merge<'a, 'b>(
     existing: Option<&'a [u8]>,
-    operands: impl IntoIterator<Item = &'a [u8]>,
+    operands: impl IntoIterator<Item = &'b [u8]>,
 ) -> Vec<u8> {
     let (mut payload, mut rc) = existing.map_or((None, 0), decode_value_with_rc);
     for (new_payload, delta) in operands.into_iter().map(decode_value_with_rc) {
-        if payload.is_none() {
-            payload = new_payload;
-        } else if new_payload.is_some() {
-            debug_assert_eq!(payload, new_payload);
-        }
         rc += delta;
+        if rc <= 0 {
+            // if the reference count drops to 0 during a merge, "release" the current value.
+            rc = 0;
+            payload = None;
+            continue;
+        }
+        match (payload, new_payload) {
+            (Some(p), Some(n)) => debug_assert_eq!(p, n),
+            (Some(_), None) => {}
+            (None, new_payload) => payload = new_payload,
+        }
     }
-
     match rc.cmp(&0) {
-        Ordering::Less => rc.to_le_bytes().to_vec(),
-        Ordering::Equal => Vec::new(),
+        Ordering::Less | Ordering::Equal => Vec::new(), // "free" the data.
```

Comment:
Is it still possible for the rc to get negative in the full merge? I'm a bit scared to approve this as I don't understand why was this implemented the current way in the first place, with the negative `rc`s.

Reply:
I believe that should not happen (especially because of the condition above that sets `rc = 0` if its negative), but `cmp` needs us to handle `Less` case somehow. We might as well put an assert in here, but that would only serve to crash the node when we least want it to.